### ISSUE TITLE
Implement the clojure.test protocol

### ISSUE THREAD

**Issue by seancorfield** [clojure-expectations/expectations#82]
This is migrated from #70 which went off-topic.

@bbatsov created the issue originally:

Basically this would allow tools like cider-test to run expectations tests, which would benefit a lot of Clojure developers. More details - https://github.com/clojure-emacs/cider#using-cider-test-with-alternative-test-libraries and clojure-emacs/cider#1606

**Comment by seancorfield** [comment]
I have a very rough proof of concept working locally but there are some nasty gotchas:

The main issue is that `(expect e a)` produces a "random" symbol name (based on a hash of the form itself): if you change the test namespace and reload it, old definitions hang around (and the changed tests get new symbol names). Expectations mitigates this by not running expectations that have `::run` metadata (which is added as each expectation completes, using Expectations custom `report` function). The "downside" is that you have to reload the test namespace, or re-evaluate the forms, to get those symbols back in play. The Emacs / REPL workflow we've adopted with Expectations is to evaluate either a single test or the whole namespace whenever we want to run them: this is kind of nice for working with tests as you write them but isn't an entirely smooth workflow overall.

If I make the CIDER integration follow that behavior, `C-c C-t r` doesn't work and neither do repeated `C-c C-t n` runs. You have to go back to the test namespace and reload it. If it doesn't follow that behavior, old tests just hang around and get re-run every time.

The other issue is that Expectations assumes a lot of context: when you run tests, you generally do it at the namespace level and it loads an `expectations-options` namespace, sets up `before-run` and `in-context` fixtures, watches for state changes etc, runs a set of tests and then tears it all down. As far as I can tell, working with CIDER test, there's no effective way to make this happen -- short of running `test-vars` for each `expect` form (i.e., for the `:test` function attached to the symbol it generates -- which will do all of the set up and tear down for each expectation).

My first approach was to do exactly that -- generate a function that ran `test-vars` and attach it as `:test` metadata but that is problematic due to setup/teardown (and it would need to rebind how the expectation reported its result which gets ugly pretty fast).

My second approach was to generate a function for `:test` that called a variant of `doexpect` that in turn called into `clojure.test/do-report` after mapping Expectations' `compare-expr` result data structure. This was much easier but loses `expectations-options` etc and still falls foul of the "random" symbol names.

I'm open to other suggestions but I think at this point the only option is to look into a syntax variant that provides fixed symbol names for expectations. This will alleviate the "random" symbol problem (and would make the existing Emacs / REPL workflow a lot smoother since expectations would no longer be "removed" as they are run, so you wouldn't need to reload the tests all the time).

If `clojure.test` compatibility -- and therefore CIDER test compatibility -- is handled via a separate namespace that relies on _named_ expectations, then much of this becomes much more tractable. Running `test-vars` on a per-symbol basis becomes much easier (even tho' it's still really the wrong way to do things) since the defined expectation function could call the `clojure.test/do-report` function instead of `expectations/report` and so on. This is going to be the path I go down next.

**Comment by seancorfield** [comment]
(copied from #70)

Regarding this particular issue of CIDER integration, I'm just not sure it can be made to work effectively without moving to a "named symbol" approach since so much of what CIDER (and the clojure.test infrastructure) assumes, is based around being able to run individual tests, so "the same syntax" would have to move a little, at least for folks who want that integration (and, FWIW, I do find naming tests, or groups of tests, to be useful from a documentation p.o.v.).

It could be as minimal as a wrapper that assigns a symbol name to one or more expectations:
```clj
(defx my-tests
  (expect e1 (some-expression v1))
  (expect e2 (some-expression v2)))
```
That would be the easiest syntax to migrate to (since you could wrap a whole namespace of expectations in a single `defx` to start with). Or it could be an additional `expect` form like this:
```clj
(expect-named my-test1 e1 (some-expression v1))
(expect-named my-test2 e2 (some-expression v2))
```
(the disadvantage is you'd need to explicitly edit and name every `expect` form in your current test suite for this).

**Comment by seancorfield** [comment]
A requirement from #85 is that we should be able to add (arbitrary) metadata to the named symbol that identifies the test, to better support the needs of tooling.

**Comment by seancorfield** [comment]
`clojure.test` tooling compatibility has now become a priority at World Singles so this is going to get worked on, starting this week. My current thinking is a new `expectations.clojure.test` namespace offering `defexpect` as a named expectation -- the key thing needed for `clojure.test` "compatibility":

    (defexpect my-test1 e1 (some-expression v1))
    (defexpect my-test2 (more-of {:key [a b]} true a coll? b) (some-expression v2)))

In other words, exactly like `expect` but introducing a fixed name for the underlying function.

The current `expectations` namespace will continue to function exactly as-is. Wherever I need to change the behavior to satisfy tooling's expectations(!) of `clojure.test`-like behavior, I will duplicate code into the new namespace and modify its behavior (most likely around the `clojure.test` reporting logic) but I'm hoping to reuse most of the comparison machinery. `defexpect` will create functions the same way `clojure.test/deftest` does: with the actual test attached as `:test` metadata and a body that calls `clojure.test/test-var` on itself.

I'll update this ticket with any issues I run into or any questions about changes in behavior that may be required for this approach.

**Comment by bbatsov** [comment]
> clojure.test tooling compatibility has now become a priority at World Singles so this is going to get worked on, starting this week.

Great news!

> The current expectations namespace will continue to function exactly as-is. Wherever I need to change the behavior to satisfy tooling's expectations(!) of clojure.test-like behavior, I will duplicate code into the new namespace and modify its behavior (most likely around the clojure.test reporting logic) but I'm hoping to reuse most of the comparison machinery. defexpect will create functions the same way clojure.test/deftest does: with the actual test attached as :test metadata and a body that calls clojure.test/test-var on itself.

Sounds like a pretty reasonable approach to me.

**Comment by seancorfield** [comment]
I have a "first cut" of this but in order to push a build to Clojars, I need to be added to the `expectations` group on clojars.org @jaycfields

**Comment by seancorfield** [comment]
If anyone wants to play with this locally, the only addition is `expectations.clojure.test/defexpect` which is a macro that names expectations.

You can either wrap a group of existing expectations:

    (ns my-example
      (:require [expectations :refer :all]
                [expectations.clojure.test :refer [defexpect]]))

    (defexpect my-tests
      (expect e1 (some-expression v1))
      (expect (more-of {:keys [a b]} true a coll? b) (some-expression v2)))

or you can just give names to existing expectations by replacing `expect` with `defexpect some-name`:

    (ns my-example
      (:require [expectations :refer :all]
                [expectations.clojure.test :refer [defexpect]]))

    (defexpect my-test1 e1 (some-expression v1))
    (defexpect my-test2 (more-of {:keys [a b]} true a coll? b) (some-expression v2)))

This creates a `deftest` behind the scenes (and turns off run-on-shutdown!) so, in theory, all `clojure.test` tooling and functionality should work with it exactly "as-is".

What you lose is some of the fancier aspects of failure reporting (and if you wrap `defexpect` around a large block of existing expectations, any failures will be reported again the line that `defexpect` is on), but it "works" as a starting point.

**Comment by seancorfield** [comment]
Note: wrapping existing expectations that have only `(expect (some-expression v3))` -- i.e., the implicit "expect true" -- doesn't work (and probably fails in a bad way).

Also, as an artifact of how the `defexpect` macro works, you can also do this:

    (ns my-example
      (:require [expectations :refer :all]
                [expectations.clojure.test :refer [defexpect]]))

    (defexpect my-tests
      e1 (some-expression v1)
      (more-of {:keys [a b]} true a coll? b) (some-expression v2))

and you can mix and match that style with `(expect e a)` but I don't think I'd recommend either of those... and it may be worth preventing that for release.

**Comment by jaycfields** [comment]
I think you're good to go @seancorfield. I added you to the clojars group.

**Comment by seancorfield** [comment]
Thanks @jaycfields -- I published a 2.2.0-SNAPSHOT to Clojars and updated the docs with an experimental clojure.test page added.

**Comment by seancorfield** [comment]
Verified `lein test` and `boot test` (via `adzerk/boot-test`) both work. Also verified the "run tests" functionality in ProtoREPL (Atom) works.

Note: the new style tests run only under `clojure.test` tooling; the old style tests run only under Expectations tooling. In order to migrate piecemeal, you'd need to run `lein do expectations, test` or `boot expectations test` (or run a watcher for each type of test).

At this stage, I'll consider this issue resolved and any bugs or enhancement requests can be raised as new issues. I'll continue to flesh out the documentation over the next week or so.
